zipline/data/data_portal.py

```diff
@@ -0,0 +1,1186 @@
+        # palladium).
+        #
+        # In both cases, we break up the dataframe into individual dfs
+        # that only contain a single asset's information.  ie, if source_df
+        # has data for PALLADIUM and GOLD, we split source_df into two
+        # dataframes, one for each. (same applies if source_df has data for
+        # AAPL and IBM).
+        #
+        # We then take each child df and reindex it to the simulation's date
+        # range by forward-filling missing values. this makes reads simpler.
+        #
+        # Finally, we store the data. For each fetcher column, we store a
+        # mapping in self.augmented_sources_map from it to a dictionary of
+        # asset -> df.  In other words,
+        # self.augmented_sources_map['days_to_cover']['AAPL'] gives us the df
+        # holding that data.
+
+        if self._sim_params.emission_rate == "daily":
+            fetcher_date_index = self.env.days_in_range(
+                start=self._sim_params.period_start,
+                end=self._sim_params.period_end
+            )
+        else:
+            fetcher_date_index = self.env.minutes_for_days_in_range(
+                start=self._sim_params.period_start,
+                end=self._sim_params.period_end
+            )
+
+        # break the source_df up into one dataframe per sid.  this lets
+        # us (more easily) calculate accurate start/end dates for each sid,
+        # de-dup data, and expand the data to fit the backtest start/end date.
+        grouped_by_sid = source_df.groupby(["sid"])
+        group_names = grouped_by_sid.groups.keys()
+        group_dict = {}
+        for group_name in group_names:
+            group_dict[group_name] = grouped_by_sid.get_group(group_name)
+
+        for identifier, df in group_dict.iteritems():
+            # before reindexing, save the earliest and latest dates
+            earliest_date = df.index[0]
+            latest_date = df.index[-1]
+
+            # since we know this df only contains a single sid, we can safely
+            # de-dupe by the index (dt)
+            df = df.groupby(level=0).last()
+
+            # reindex the dataframe based on the backtest start/end date.
+            # this makes reads easier during the backtest.
+            df = df.reindex(index=fetcher_date_index, method='ffill')
+
+            if not isinstance(identifier, Asset):
+                # for fake assets we need to store a start/end date
+                self._asset_start_dates[identifier] = earliest_date
+                self._asset_end_dates[identifier] = latest_date
+
+            for col_name in df.columns.difference(['sid']):
+                if col_name not in self._augmented_sources_map:
+                    self._augmented_sources_map[col_name] = {}
+
+                self._augmented_sources_map[col_name][identifier] = df
+
+    def _open_daily_file(self):
+        if self._daily_equities_data is None:
```

Comment:
Not sure yet whether these methods fall in inner loops or not, but I notice that on every call they check state that isn't  changing between calls, unless the method itself changes that state.  The `is None` checks are probably negligible, but I'm not sure about the ones like `self.data_frequency == 'daily'`.  We could look into doing them just once per algo run.